app/tx_cache.go

```diff
@@ -0,0 +1,55 @@
+package app
+
+import (
+	"crypto/sha256"
+	"sync"
+)
+
+// TxValidationCache caches the results of expensive transaction validation
+// to avoid repeating the same work in ProcessProposal that was already done in CheckTx.
+type TxValidationCache struct {
+	cache sync.Map
+}
+
+// NewTxValidationCache creates a new transaction validation cache
+func NewTxValidationCache() *TxValidationCache {
+	return &TxValidationCache{
+		cache: sync.Map{},
+	}
+}
+
+// getTxKey generates a deterministic key for a transaction
+func (c *TxValidationCache) getTxKey(tx []byte) string {
+	hash := sha256.Sum256(tx)
+	return string(hash[:])
+}
+
+// Exists retrieves a validation result from the cache
+func (c *TxValidationCache) Exists(tx []byte) (exists bool) {
+	key := c.getTxKey(tx)
+	_, exists = c.cache.Load(key)
+	return exists
+}
+
+// Set stores a validation result in the cache
+func (c *TxValidationCache) Set(tx []byte) {
+	key := c.getTxKey(tx)
+	c.cache.Store(key, struct{}{})
+}
```

Comment:
[question] why isn't the value in the map a `bool` or some other type that indicates the result of validation?